File: VB/TimeSpanEditor/TimeSpanEditor.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors
Imports System.ComponentModel
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.Utils
Imports DevExpress.Data.Mask
Imports System.Globalization
Imports DevExpress.XtraEditors.Mask

Namespace TimeSpanEditor
	<UserRepositoryItem("RegisterTimeSpanEdit")> _
	Public Class RepositoryItemTimeSpanEdit
		Inherits RepositoryItemTimeEdit
		Private allowDayInput_Renamed As Boolean
		Shared Sub New()
			RegisterTimeSpanEdit()
		End Sub
		Public Sub New()
			allowDayInput_Renamed = False
			UpdateFormats()
		End Sub
		Public Const TimeSpanEditName As String = "TimeSpanEdit"
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return TimeSpanEditName
			End Get
		End Property
		Public Shared Sub RegisterTimeSpanEdit()
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(TimeSpanEditName, GetType(TimeSpanEdit), GetType(RepositoryItemTimeSpanEdit), GetType(BaseSpinEditViewInfo), New ButtonEditPainter(), True))
		End Sub
		<Browsable(False)> _
		Public Overrides ReadOnly Property EditFormat() As FormatInfo
			Get
				Return MyBase.EditFormat
			End Get
		End Property
		<Browsable(False)> _
		Public Overrides ReadOnly Property DisplayFormat() As FormatInfo
			Get
				Return MyBase.DisplayFormat
			End Get
		End Property
		<Browsable(False)> _
		Public Overrides ReadOnly Property Mask() As MaskProperties
			Get
				Return MyBase.Mask
			End Get
		End Property
		<Browsable(False)> _
		Public Shadows Overridable ReadOnly Property EditMask() As String
			Get
				Dim mask As String = "HH:mm:ss"
				If AllowDayInput Then
					mask = "d." & mask
				End If
				Return mask
			End Get
		End Property
		Protected Friend Overridable ReadOnly Property TimeSeparator() As Char
			Get
				Return TimeSpanHelper.TimeSeparator
			End Get
		End Property
		Protected Friend Overridable ReadOnly Property DaySeparator() As Char
			Get
				Return TimeSpanHelper.DaySeparator
			End Get
		End Property
		<Category(CategoryName.Behavior), DefaultValue(False)> _
		Public Overridable Property AllowDayInput() As Boolean
			Get
				Return allowDayInput_Renamed
			End Get
			Set(ByVal value As Boolean)
				If allowDayInput_Renamed = value Then
					Return
				End If
				allowDayInput_Renamed = value
				UpdateFormats()
			End Set
		End Property
		Protected Overridable Sub UpdateFormats()
			EditFormat.FormatString = EditMask
			DisplayFormat.FormatString = EditMask
			Mask.EditMask = EditMask
		End Sub
		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			BeginUpdate()
			Try
				MyBase.Assign(item)
				Dim source As RepositoryItemTimeSpanEdit = TryCast(item, RepositoryItemTimeSpanEdit)
				If source Is Nothing Then
					Return
				End If
				Me.AllowDayInput = source.AllowDayInput
			Finally
				EndUpdate()
			End Try
		End Sub

		Public Overrides Function GetDisplayText(ByVal format As FormatInfo, ByVal editValue As Object) As String
			If TypeOf editValue Is TimeSpan Then
				Return TimeSpanHelper.TimeSpanToString((CType(editValue, TimeSpan)),AllowDayInput)
			End If
			If TypeOf editValue Is String Then
				Return editValue.ToString()
			End If
			Return GetDisplayText(Nothing, New TimeSpan(0))
		End Function

		Protected Friend Overridable Function GetFormatMaskAccessFunction(ByVal editMask As String, ByVal managerCultureInfo As CultureInfo) As String
			Return GetFormatMask(editMask, managerCultureInfo)
		End Function
	End Class

	Public Class TimeSpanEdit
		Inherits TimeEdit
		Shared Sub New()
			RepositoryItemTimeSpanEdit.RegisterTimeSpanEdit()
		End Sub
		Public Sub New()
			MyBase.New()
			Me.fEditValue = New TimeSpan(0)
			Me.fOldEditValue = Me.fEditValue
		End Sub
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemTimeSpanEdit.TimeSpanEditName
			End Get
		End Property
		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As RepositoryItemTimeSpanEdit
			Get
				Return TryCast(MyBase.Properties, RepositoryItemTimeSpanEdit)
			End Get
		End Property
		Public Overrides Property EditValue() As Object
			Get
				If Properties.ExportMode = ExportMode.DisplayText Then
					Return Properties.GetDisplayText(Nothing, MyBase.EditValue)
				End If
				Return MyBase.EditValue
			End Get
			Set(ByVal value As Object)
				If TypeOf value Is DateTime Then
					Dim time As DateTime = (CDate(value))
					MyBase.EditValue = New TimeSpan(time.Ticks)

				ElseIf TypeOf value Is TimeSpan Then
					MyBase.EditValue = value
				ElseIf TypeOf value Is String Then
					MyBase.EditValue = TimeSpanHelper.Parse(CStr(value))
				Else
					MyBase.EditValue = New TimeSpan(0, 0, 0)
				End If
			End Set
		End Property


		Protected Overrides Function CreateMaskManager(ByVal mask As MaskProperties) As MaskManager
			Dim patchedMask As New CustomTimeEditMaskProperties()
			patchedMask.Assign(mask)
			patchedMask.EditMask = Properties.GetFormatMaskAccessFunction(mask.EditMask, mask.Culture)
			Return patchedMask.CreatePatchedMaskManager()
		End Function
	End Class

	Public Class CustomTimeEditMaskProperties
		Inherits TimeEditMaskProperties
		Public Sub New()
			MyBase.New()
		End Sub
		Public Overridable Function CreatePatchedMaskManager() As MaskManager
			Dim managerCultureInfo As CultureInfo = Me.Culture
			If managerCultureInfo Is Nothing Then
				managerCultureInfo = CultureInfo.CurrentCulture
			End If
			Dim editMask As String = Me.EditMask
			If editMask Is Nothing Then
				editMask = String.Empty
			End If
			Return New CustomDateTimeMaskManager(editMask, False, managerCultureInfo, True)
		End Function
	End Class



End Namespace

File: CS/TimeSpanEditor/TimeSpanEditor.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors;
using System.ComponentModel;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using DevExpress.Utils;
using DevExpress.Data.Mask;
using System.Globalization;
using DevExpress.XtraEditors.Mask;

namespace TimeSpanEditor
{
    [UserRepositoryItem("RegisterTimeSpanEdit")]
    public class RepositoryItemTimeSpanEdit : RepositoryItemTimeEdit
    {
        bool allowDayInput;
        static RepositoryItemTimeSpanEdit() { RegisterTimeSpanEdit(); }
        public RepositoryItemTimeSpanEdit() {
            allowDayInput = false;
            UpdateFormats();
        }
        public const string TimeSpanEditName = "TimeSpanEdit";
        public override string EditorTypeName { get { return TimeSpanEditName; } }
        public static void RegisterTimeSpanEdit()
        {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(TimeSpanEditName,
              typeof(TimeSpanEdit), typeof(RepositoryItemTimeSpanEdit),
              typeof(BaseSpinEditViewInfo), new ButtonEditPainter(), true));
        }
        [Browsable(false)]
        public override FormatInfo EditFormat { get { return base.EditFormat; } }
        [Browsable(false)]
        public override FormatInfo DisplayFormat { get { return base.DisplayFormat; } }
        [Browsable(false)]
        public override MaskProperties Mask { get { return base.Mask; } }
        [Browsable(false)]
        public new virtual string EditMask { 
            get {
                string mask = "HH:mm:ss";
                if (AllowDayInput) mask = "d." + mask;
                return mask; 
            } 
        }
        protected internal virtual char TimeSeparator { get { return TimeSpanHelper.TimeSeparator; } }
        protected internal virtual char DaySeparator { get { return TimeSpanHelper.DaySeparator; } }
        [Category(CategoryName.Behavior), DefaultValue(false)]
        public virtual bool AllowDayInput 
        {
            get { return allowDayInput; }
            set { 
                if (allowDayInput == value) return;
                allowDayInput = value;
                UpdateFormats();
            }
        }
        protected virtual void UpdateFormats() {
            EditFormat.FormatString = EditMask;
            DisplayFormat.FormatString = EditMask;
            Mask.EditMask = EditMask;
        }
        public override void Assign(RepositoryItem item)
        {
            BeginUpdate();
            try
            {
                base.Assign(item);
                RepositoryItemTimeSpanEdit source = item as RepositoryItemTimeSpanEdit;
                if (source == null) return;
                this.AllowDayInput = source.AllowDayInput;
            }
            finally
            {
                EndUpdate();
            }
        }

        public override string GetDisplayText(FormatInfo format, object editValue)
        {
            if (editValue is TimeSpan)
                return TimeSpanHelper.TimeSpanToString(((TimeSpan)editValue),AllowDayInput );
            if (editValue is string)
                return editValue.ToString();
            return GetDisplayText(null, new TimeSpan(0));
        }

        protected internal virtual string GetFormatMaskAccessFunction(string editMask, CultureInfo managerCultureInfo)
        {
            return GetFormatMask(editMask, managerCultureInfo);
        }
    }

    public class TimeSpanEdit : TimeEdit
    {
        static TimeSpanEdit() { RepositoryItemTimeSpanEdit.RegisterTimeSpanEdit(); }
        public TimeSpanEdit()
            : base()
        {
            this.fOldEditValue = this.fEditValue = new TimeSpan(0);
        }
        public override string EditorTypeName { get { return RepositoryItemTimeSpanEdit.TimeSpanEditName; } }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemTimeSpanEdit Properties { get { return base.Properties as RepositoryItemTimeSpanEdit; } }
        public override object EditValue
        {
            get
            {
                if (Properties.ExportMode == ExportMode.DisplayText)
                    return Properties.GetDisplayText(null, base.EditValue); 
                return base.EditValue;
            }
            set
            {
                if (value is DateTime)
                {
                    DateTime time = ((DateTime)value);
                    base.EditValue = new TimeSpan(time.Ticks);

                }
                else if (value is TimeSpan)
                    base.EditValue = value;
                else if (value is string)
                    base.EditValue = TimeSpanHelper.Parse((string)value);
                else
                    base.EditValue = new TimeSpan(0, 0, 0);
            }
        }


        protected override MaskManager CreateMaskManager(MaskProperties mask)
        {
            CustomTimeEditMaskProperties patchedMask = new CustomTimeEditMaskProperties();
            patchedMask.Assign(mask);
            patchedMask.EditMask = Properties.GetFormatMaskAccessFunction(mask.EditMask, mask.Culture);
            return patchedMask.CreatePatchedMaskManager();
        }
    }
    
    public class CustomTimeEditMaskProperties : TimeEditMaskProperties
    {
        public CustomTimeEditMaskProperties() : base() { }
        public virtual MaskManager CreatePatchedMaskManager()
        {
            CultureInfo managerCultureInfo = this.Culture;
            if (managerCultureInfo == null)
                managerCultureInfo = CultureInfo.CurrentCulture;
            string editMask = this.EditMask;
            if (editMask == null)
                editMask = string.Empty;
            return new CustomDateTimeMaskManager(editMask, false, managerCultureInfo, true);
        }
    }



}
